operators/src/engine/query.rs

```diff
@@ -43,37 +51,96 @@ impl TestDefault for ChunkByteSize {
 pub trait QueryContext: Send + Sync {
     fn chunk_byte_size(&self) -> ChunkByteSize;
     fn thread_pool(&self) -> &Arc<ThreadPool>;
+
+    fn abort_registration(&self) -> &QueryAbortRegistration;
+    fn abort_trigger(&mut self) -> Result<QueryAbortTrigger>;
+}
+
+pub struct QueryAbortRegistration {
+    valve: Valve,
+}
+
+impl QueryAbortRegistration {
+    pub fn new() -> (Self, QueryAbortTrigger) {
+        let (trigger, valve) = Valve::new();
+
+        (Self { valve }, QueryAbortTrigger { trigger })
+    }
+
+    pub fn wrap<S: Stream>(&self, stream: S) -> QueryAbortWrapper<S> {
+        QueryAbortWrapper {
+            valved: self.valve.wrap(stream),
+        }
+    }
+}
+
+#[pin_project(project = AbortWrapperProjection)]
+pub struct QueryAbortWrapper<S> {
+    #[pin]
+    valved: Valved<S>,
+}
+
+impl<S> Stream for QueryAbortWrapper<S>
+where
+    S: Stream,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().valved.poll_next(cx)
+    }
+}
```

Comment:
if we would use `AbortableStream` here, we would just need to `poll_next` on poll next and to call abort on "`valved`" when abort is called, right? We would store the `AbortHandle` as a second field in `QueryAbortWrapper`.
  Shouldn't we do this then? Couldn't we omit the registration then?

Reply:
The `futures::stream::Abortable` doc says

> When abort is called on the handle tied to reg or if abort has already been called, the future/stream will complete immediately without making any further progress.

so if it completes immediately we have no chance to call abort on the other streams, right? Also, we still would have to have the `abort_handles` of all the other wrapped streams. And we can't collect them in the `QueryContext` since it can't be `mut`.

Reply:
1. The thing about … being mutable is strange. How do you think you can add something to the current struct? It is also interior mutability.

2. If you create a stream wrapper. What is the difference in adding 
 a. an atomic bool to the stream that says: If stopped then return X; else poll_next  – and you share this bool for all streams in the workflow?
 b. using https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.take_until ?

I'm still curious if this could be simplified. Currently, you create two Streams for each stream, right?

Reply:
> The thing about … being mutable is strange. How do you think you can add something to the current struct? It is also interior mutability.

No, it works with a channel or something.

> a. an atomic bool to the stream that says: If stopped then return X; else poll_next – and you share this bool for all streams in the workflow?

doesn't this require locking on each poll?

> b. using https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.take_until ?

stream-cancel uses something very similar I think

> I'm still curious if this could be simplified. Currently, you create two Streams for each stream, right?

every processor creates a stream (logic) and I wrap it (to make it abortable). So it's one stream.

Reply:
> I guess not, since its an atomic load: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.load

ok it's a lock on the hardware level then

> Ah okay, so in principle we could use a single produce, multi receiver channel where the receiver closes on the first message.

yeah, stream-cancel uses something called a `TripWire`, built upon a `tokio::sync::watch` Channel. That's basically a Future on which multiple threads can poll at the same time. So we could set up something like this ourselves but we'd have to do that and test it. Which was already done in the stream-cancel trait.

> Okay, since valved (which also creates a stream) is re-used?

no it's not reused. All the valves only wait on the same Future in the `take_until`.